src/source/src/table.rs

```diff
@@ -89,34 +88,42 @@ impl TableDmlHandle {
             // tasks to the compute nodes, so this'll be temporarily unavailable, so we throw an
             // error instead of asserting here.
             // TODO: may reject DML when streaming executors are not recovered.
-            let tx = self
-                .core
-                .read()
-                .changes_txs
-                .iter()
-                .choose(&mut rand::thread_rng())
-                .context("no available table reader in streaming source executors")?
-                .clone();
+            let tx = {
+                let guard = self.core.read();
+                if guard.changes_txs.is_empty() {
+                    return Err(RwError::from(anyhow!(
+                        "no available table reader in streaming source executors"
+                    )));
+                }
+                let len = guard.changes_txs.len();
+                guard
+                    .changes_txs
+                    .get((txn_msg.txn_id() % len as u64) as usize)
+                    .context("no available table reader in streaming source executors")?
+                    .clone()
+            };
 
             #[cfg(debug_assertions)]
-            risingwave_common::util::schema_check::schema_check(
-                self.column_descs
-                    .iter()
-                    .filter_map(|c| (!c.is_generated()).then_some(&c.data_type)),
-                chunk.columns(),
-            )
-            .expect("table source write chunk schema check failed");
+            if let TxnMsg::Data(_, chunk) = &txn_msg {
+                risingwave_common::util::schema_check::schema_check(
+                    self.column_descs
+                        .iter()
+                        .filter_map(|c| (!c.is_generated()).then_some(&c.data_type)),
+                    chunk.columns(),
+                )
+                .expect("table source write chunk schema check failed");
+            }
 
             let (notifier_tx, notifier_rx) = oneshot::channel();
 
-            match tx.send((chunk, notifier_tx)).await {
+            match tx.send((txn_msg, notifier_tx)).await {
                 Ok(_) => return Ok(notifier_rx),
 
                 // It's possible that the source executor is scaled in or migrated, so the channel
                 // is closed. In this case, we should remove the closed channel and retry.
```

Comment:
In this case, it seems still possible that the DML chunks are sent to multiple channels. 🤔

Reply:
It seems the scale of source executors could even cause the termination of DMLs. Besides, altering tables can cause the same problem. I don't have a good idea to handle them right now.

Reply:
BTW, once we can resolve this problem, there'll be no need to collect notifiers to a `Vec`. Only waiting for the last write is okay.